chalk-ir/src/visit/visitors.rs

```diff
@@ -0,0 +1,71 @@
+use crate::{BoundVar, DebruijnIndex, Interner, Visit, VisitResult, Visitor};
+
+pub trait VisitExt<I: Interner>: Visit<I> {
+    fn has_free_vars(&self, interner: &I) -> bool {
+        self.visit_with(
+            &mut FindFreeVarsVisitor { interner },
+            DebruijnIndex::INNERMOST,
+        )
+        .to_bool()
+    }
+}
+
+impl<T, I: Interner> VisitExt<I> for T where T: Visit<I> {}
+
+#[derive(Clone, Copy, Debug)]
+pub struct FindAny {
+    pub found: bool,
+}
+
+impl FindAny {
+    pub fn to_bool(&self) -> bool {
+        self.found
+    }
+}
+
+impl VisitResult for FindAny {
+    fn new() -> Self {
+        FindAny { found: false }
+    }
+
+    fn return_early(&self) -> bool {
+        self.found
+    }
+    fn combine(self, other: Self) -> Self {
+        FindAny {
+            found: self.found || other.found,
+        }
+    }
+}
+
+pub struct FindFreeVarsVisitor<'i, I: Interner> {
```

Comment:
I feel like this should be a private impl detail of the `has_free_vars` function